blockchain/message_pool/src/msgpool/selection.rs

```diff
@@ -0,0 +1,592 @@
+            last = i;
+            break;
+        }
+        // 4. We have reached the edge of what we can fit wholesale; if we still have available gasLimit
+        // to pack some more chains, then trim the last chain and push it down.
+        // Trimming invalidates subsequent dependent chains so that they can't be selected as their
+        // dependency cannot be (fully) included.
+        // We do this in a loop because the blocker might have been inordinately large and we might
+        // have to do it multiple times to satisfy tail packing.
+        'tail_loop: while gas_limit >= min_gas && last < chains.len() {
+            chains[last].trim(gas_limit, &base_fee, allow_negative_chains(cur_ts.epoch()));
+            if chains[last].curr().valid {
+                for i in last..(chains.len() - 1) {
+                    if chains[i].compare(&chains[i + 1]) == Ordering::Greater {
+                        break;
+                    }
+                    chains.swap(i, i + 1);
+                }
+            }
+
+            // select the next (valid and fitting) chain for inclusion
+            for i in last..chains.len() {
+                if !chains[i].curr().valid {
+                    continue;
+                }
+
+                // if gas_perf < 0 then we have no more profitable chains
+                if !allow_negative_chains(cur_ts.epoch()) && chains[i].curr().gas_perf < 0.0 {
+                    break 'tail_loop;
+                }
+
+                // does it fit in the block?
+                if chains[i].curr().gas_limit <= gas_limit {
+                    gas_limit -= chains[i].curr().gas_limit;
+                    result.append(&mut chains[i].curr_mut().msgs);
+                    continue;
+                }
+                last += i;
+                continue 'tail_loop;
+            }
+            break;
+        }
+        Ok(result)
+    }
+
+    async fn get_pending_messages(&self, cur_ts: &Tipset, ts: &Tipset) -> Result<Pending, Error> {
+        let mut result: Pending = HashMap::new();
+        let mut in_sync = false;
+        if cur_ts.epoch() == ts.epoch() && cur_ts == ts {
+            in_sync = true;
+        }
+
+        for (a, mset) in self.pending.read().await.iter() {
+            if in_sync {
+                result.insert(*a, mset.msgs.clone());
+            } else {
+                let mut mset_copy = HashMap::new();
+                for (nonce, m) in mset.msgs.iter() {
+                    mset_copy.insert(*nonce, m.clone());
+                }
+                result.insert(*a, mset_copy);
+            }
+        }
+
+        if in_sync {
+            return Ok(result);
+        }
+
+        // Run head change to do reorg detection
+        run_head_change(
+            &self.api,
+            &self.pending,
+            cur_ts.clone(),
+            ts.clone(),
+            &mut result,
+        )
+        .await?;
+
+        Ok(result)
+    }
+
+    async fn select_priority_messages(
```

Comment:
This seems to have a ton of overlap with `select_messages_greedy`, can we do better?

Reply:
I fine its a little easier (for me) to read like this weirdly, but yeah makes sense, i can remove some of the over lap. 👍